pkg/gardenlet/controller/shoot/shoot_control.go

```diff
@@ -837,6 +837,47 @@ func computeOperationType(shoot *gardencorev1beta1.Shoot) gardencorev1beta1.Last
 	return v1beta1helper.ComputeOperationType(shoot.ObjectMeta, shoot.Status.LastOperation)
 }
 
+func needsControlPlaneDeployment(ctx context.Context, o *operation.Operation, kubeAPIServerDeploymentFound bool, infrastructure *extensionsv1alpha1.Infrastructure) (bool, error) {
+	var (
+		namespace = o.Shoot.SeedNamespace
+		name      = o.Shoot.GetInfo().Name
+	)
+
+	// If the `ControlPlane` resource and the kube-apiserver deployment do no longer exist then we don't want to re-deploy it.
+	// The reason for the second condition is that some providers inject a cloud-provider-config into the kube-apiserver deployment
+	// which is needed for it to run.
+	exists, markedForDeletion, err := extensionResourceStillExists(ctx, o.K8sSeedClient.APIReader(), &extensionsv1alpha1.ControlPlane{}, namespace, name)
+	if err != nil {
+		return false, err
+	}
+
+	switch {
+	// treat `ControlPlane` in deletion as if it is already gone. If it is marked for deletion, we also shouldn't wait
+	// for it to be reconciled, as it can potentially block the whole deletion flow (deletion depends on other control
+	// plane components like kcm and grm) which are scaled up later in the flow
+	case (!exists || markedForDeletion) && !kubeAPIServerDeploymentFound:
```

Comment:
Earlier, this was

```golang
	if !exists && !kubeAPIServerDeploymentFound || markedForDeletion {
```

is this change intended?

Reply:
Nice catch. I must have reverted it to the older logic by accident when I was rebasing the PR.